src/conn/conn_layered.c

```diff
@@ -1874,158 +1869,191 @@ __layered_update_prune_timestamps_print_update_logs(WT_SESSION_IMPL *session,
+}
 
-            /*
-             * We now have the oldest checkpoint in use for this table. If it's different from the
-             * last time we checked, update the timestamp for pruning.
-             */
-            if (ckpt_inuse != layered_table->last_ckpt_inuse) {
-                /*
-                 * Set the prune timestamp in the btree if it is open, typically it is. However,
-                 * it's possible that it hasn't been opened yet. In that case, we need to skip
-                 * updating its timestamp for pruning, and we'll get another chance to update the
-                 * prune timestamp at the next checkpoint.
-                 */
-                WT_ERR_NOTFOUND_OK(
-                  __wt_session_get_dhandle(session, layered_table->ingest_uri, NULL, NULL, 0),
-                  true);
-                if (ret != WT_NOTFOUND) {
-                    btree = (WT_BTREE *)session->dhandle->handle;
-
-                    __layered_update_prune_timestamps_print_update_logs(
-                      session, layered_table, prune_timestamp, ckpt_inuse);
-
-                    WT_ASSERT(session, prune_timestamp >= btree->prune_timestamp);
-                    __wt_atomic_store_uint64_release(&btree->prune_timestamp, prune_timestamp);
-                    layered_table->last_ckpt_inuse = ckpt_inuse;
-
-                    WT_ERR(__wt_session_release_dhandle(session));
-                } else {
-                    __wt_verbose_level(session, WT_VERB_LAYERED, WT_VERBOSE_DEBUG_5,
-                      "GC %s: Handle not found for ingest table uri: %s", layered_table->iface.name,
-                      layered_table->ingest_uri);
-                    ret = 0;
-                }
-            }
-        }
+/*
+ * __layered_update_gc_ingest_tables_prune_timestamps --
+ *     Iterate over all ingest tables and check whether their prune timestamps could be updated.
+ */
+static int
+__layered_update_gc_ingest_tables_prune_timestamps(
+  WT_SESSION_IMPL *session, wt_timestamp_t checkpoint_timestamp)
+{
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_ITEM(layered_table_uri_buf);
+    WT_DECL_ITEM(uri_at_checkpoint_buf);
+    WT_DECL_RET;
+    WT_LAYERED_TABLE_MANAGER *manager;
+    WT_LAYERED_TABLE_MANAGER_ENTRY *entry;
+    size_t i;
+
+    conn = S2C(session);
+    manager = &conn->layered_table_manager;
+    WT_RET(__wt_scr_alloc(session, 0, &layered_table_uri_buf));
+    WT_RET(__wt_scr_alloc(session, 0, &uri_at_checkpoint_buf));
+
+    WT_ASSERT(session, manager->init);
+
+    __wt_spin_lock(session, &manager->layered_table_lock);
+    for (i = 0; i < manager->open_layered_table_count; i++) {
+        if ((entry = manager->entries[i]) == NULL)
+            continue;
+        WT_ERR(__wt_buf_setstr(session, layered_table_uri_buf, entry->layered_uri));
+
+        /*
+         * Unlock the mutex while handling a table since while updating the prune timestamp we get a
+         * dhandle lock which could cause a deadlock.
```

Comment:
Can we add something to this comment?  I'm thinking of:  "Releasing the mutex may allow the table to grow, shrink or be modified during this operation. It's okay to prune an element twice in a loop (the second pruning will probably do nothing), or miss an element to prune (it will be visited next time)."